### PR TITLE
samples: add userspace producer/consumer sample

### LINKED ISSUE #14683: need end-to-end memory protection samples
We need an end-to-end memory protection sample which demonstrates the following. We have ample tests for these but not an example showing everything working together.

- [x] Multiple logical applications, each having their own threads, partitioned with memory domains
- [x] Instantiation and assignment to a partition of a sys_mem_pool
- [x] Configuration and use of thread resource pools
- [x] Management of permission of drivers and kernel objects
- [ ] Allocation of dynamic kernel objects
- [x] Definition of an application-specific system call
- [x] Important: show various methods of data exchange where ISRs/callbacks use IPC mechanisms to give data to user mode.

### PR DESCRIPTION
**Pull request by tyrantlizard77**
We try to demonstrate some concept for user mode:

- Multiple logical applications, each with their own memory
  domain
- Creation of a sys_mem_pool and assignment to a memory
  partition
- Use of APIs like k_queue_alloc_append() which require
  thread resource pools to be configured
- Management of permissions for kernel objects and drivers
- Show how application-specific system calls are defined
- Show IPC between ISR and application (using k_msgq) and
  application-to-application IPC (using k_queue)

Fixes: #14683

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

### CODE REVIEW

**Comment by andrewboie** [comment]
Sample execution output:

```
I: APP A partition: 0x00110000 4096
I: Shared partition: 0x0010e000 4096
I: sample_driver_foo_isr: param=0x00147078 count=0
I: monitor thread got data payload #0
I: sample_driver_foo_isr: param=0x00147078 count=1
I: monitor thread got data payload #1
I: sample_driver_foo_isr: param=0x00147078 count=2
I: monitor thread got data payload #2
I: sample_driver_foo_isr: param=0x00147078 count=3
I: monitor thread got data payload #3
I: sample_driver_foo_isr: param=0x00147078 count=4
I: monitor thread got data payload #4
I: processing payload #1 complete
I: writing processed data blob back to the sample device
I: sample_driver_foo_isr: param=0x00147078 count=5
I: monitor thread got data payload #5
I: processing payload #2 complete
I: writing processed data blob back to the sample device
I: sample_driver_foo_isr: param=0x00147078 count=6
I: monitor thread got data payload #6
I: processing payload #3 complete
I: writing processed data blob back to the sample device
I: sample_driver_foo_isr: param=0x00147078 count=7
I: monitor thread got data payload #7
I: processing payload #4 complete
I: writing processed data blob back to the sample device
I: sample_driver_foo_isr: param=0x00147078 count=8
I: monitor thread got data payload #8
I: processing payload #5 complete
I: writing processed data blob back to the sample device
I: sample_driver_foo_isr: param=0x00147078 count=9
I: monitor thread got data payload #9
I: processing payload #6 complete
I: writing processed data blob back to the sample device
I: processing payload #7 complete
I: writing processed data blob back to the sample device
I: processing payload #8 complete
I: writing processed data blob back to the sample device
I: processing payload #9 complete
I: writing processed data blob back to the sample device
I: processing payload #10 complete
I: writing processed data blob back to the sample device
I: SUCCESS
```

**Review by wentongwu — APPROVED** [review]

File: samples/userspace/prod_consumer/src/main.c [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+/*
+ * Sample application that exercises various user mode concepts
+ *
+ * Consider a "sample driver" which gets incoming data from some unknown source
+ * and generates interrupts with pointers to this data. The application needs
+ * to perform some processing on this data and then write the processed data
+ * back to the driver.
+ *
+ * In this example, we have an Application A whose job it is to talk to the
```
Comment (ioannisg) [inline samples/userspace/prod_consumer/src/main.c]: remove "it"

File: samples/userspace/prod_consumer/src/main.c [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+		  WB_UP(sizeof(struct sys_mem_pool_block)))
+SYS_MEM_POOL_DEFINE(shared_pool, NULL, BLK_SIZE, BLK_SIZE, 8, 8,
+		    K_APP_DMEM_SECTION(shared_partition));
+
+/* queues for exchanging data between App A and App B */
+K_QUEUE_DEFINE(shared_queue_incoming);
+K_QUEUE_DEFINE(shared_queue_outgoing);
+
+/* Define a thread for the root of application A.
+ * We don't do this for Applicattion B, instead we re-use the main thread
```
Comment (ioannisg) [inline samples/userspace/prod_consumer/src/main.c]: typo here

**Review by ioannisg — COMMENTED** [review]
Looks good. Tested it on all ARM MPU variants (v6-M, v7-M, v8-M and NXP). Works fine.

Left some minor doc comments.
Would be nice to have a README file where the text in main.c could be moved to.
Also, we should list an expected console output.

**Review by dcpleung — APPROVED** [review]

**Comment by andrewboie** [comment]
@ioannisg thanks for reviewing, updated.

**Review by ioannisg — APPROVED** [review]

File: samples/userspace/prod_consumer/README [hunk 3]
```diff
@@ -0,0 +1,93 @@
+Producer/consumer Sample application that exercises some user mode concepts
```
Comment (nashif) [inline samples/userspace/prod_consumer/README]: Please make this an RST file so it can be part of the documentation.

**Review by nashif — DISMISSED** [review]

File: samples/userspace/prod_consumer/src/sample_driver_foo.c [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+#include <kernel.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(sample_driver);
+
+/*
+ * Fake sample driver for demonstration purposes
+ *
+ * This is fake driver for demonstration purposes, showing how an application
+ * can make system calls to interact with it.
```
Comment (ceolin) [inline samples/userspace/prod_consumer/src/sample_driver_foo.c]: Just one minor comment, the syscall itself is not interacting with this driver, so this comment is a little bit confuse.

**Review by ceolin — APPROVED** [review]

File: samples/userspace/prod_consumer/README (hunk 3 above)
Comment (andrewboie) [reply samples/userspace/prod_consumer/README]: done

File: samples/userspace/prod_consumer/src/sample_driver_foo.c (hunk 4 above)
Comment (ceolin) [reply samples/userspace/prod_consumer/src/sample_driver_foo.c]: the application is but I understood reading it that the syscall would communicate/interact with the driver

**Pull request merged**